breakdown/mem0.md

```diff
@@ -0,0 +1,307 @@
+- **Graph Database:** Memo-Graph uses Neo4j.
+- **Hyperparameters:** `m = 10` (recent messages), `s = 10` (similar memories).
+- **Asynchronous Processing:** Summary generation runs independently for up-to-date global context.
+
+## 5. Data Structures and Algorithms
+
+## Key Data Structures
+
+### Memo - Memory Fact (ω)
+
+```python
+class MemoryFact:
+        id: str
+        content: str # e.g., "User is a vegetarian and avoids dairy."
+        timestamp: str
+        created_by_tool: str
+```
+
+### Memo-Graph - Entity Node (v)
+
+```python
+class EntityNode:
+        id: str
+        label: str # e.g., "Person", "Location", "DietaryPreference"
+        entity_type: str
+        embedding: List[float] # Semantic vector representation
+        metadata: Dict[str, any] # e.g., {"creation_timestamp": "..."}
+```
+
+### Memo-Graph - Relationship Triplet ((vs, r, vd))
+
+```python
+class RelationshipTriplet:
+        source_node: EntityNode
+        destination_node: EntityNode
+        relationship_type: str # e.g., "LIVES_IN", "PREFERS", "IS_A"
+```
+
+## Algorithms
+
+- **Memory Extraction (Memo):**
+  - Input: New message pair, summary, recent messages.
+  - Process: Concatenate inputs, prompt LLM to extract salient facts.
+  - Output: List of candidate memory facts.
+
+- **Conflict Resolution (Memo-Graph):**
+  - Input: New relationship triplet.
+  - Process: Search graph for similar nodes, identify conflicting relationships, resolve via LLM.
+  - Output: Updated, consistent knowledge graph.
+
+## Performance Considerations
+
+- **Latency:** Memo p95 latency 1.44s (91% faster than full-context). Memo-Graph p95 latency 2.59s.
+- **Token Cost:** Memo averages 7k tokens/conversation, Memo-Graph 14k, much less than full history (~26k tokens).
+- **Trade-offs:** Memo is fastest for single-hop queries; Memo-Graph excels at complex, multi-hop/temporal queries.
+
+## 6. Technical Challenges and Innovations
+
+## Problems Solved
+
+- **Context Window Limitation:** Enables long-term memory beyond LLM context window.
+- **Conversational Amnesia:** Prevents agents from forgetting preferences or repeating questions.
+- **Inefficient Context Retrieval:** Retrieves concise facts, improving LLM signal-to-noise ratio.
+- **Reasoning over Time:** Memo-Graph models event sequences for temporal reasoning.
+
+## Unique Approaches
+
+- **LLM as Database Operator:** LLM manages memory integrity (ADD/UPDATE/DELETE/NOOP).
+- **Hybrid Contextualization:** Combines global summary and recent messages for extraction.
+- **Logical Deletion in Graphs:** Marks conflicting info as invalid, preserving historical data.
+
+## Technical Breakthroughs
+
+- **State-of-the-Art Performance:** Top scores on LOCOMO benchmark, outperforming existing systems.
+- **Production-Ready Efficiency:** >90% reductions in latency and token cost.
+- **Effective Relational Modeling:** Structured graph representation improves multi-hop/temporal reasoning.
```

Comment:
This section looks too brief. It should have more explanation, esp, on the how part